Parse an ellipsis `T...` for type parameter packs

hamishknight opened the pull request:
In a generic parameter list, parse an ellipsis to produce a type parameter pack. This replaces the previous `@_typeSequence` attribute.

slavapestov commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -1964,6 +1964,13 @@ ERROR(attr_requires_concurrency, none,
+//------------------------------------------------------------------------------
+
+ERROR(associatedtype_cannot_be_pack,none,
+      "associated types may not use type parameter packs", ())
```
'associated types cannot be variadic' perhaps?

hamishknight replied:
Sounds good to me!

slavapestov reviewed (approved):
Awesome, thank you!

hamishknight committed:
Parse an ellipsis `T...` for type parameter packs

In a generic parameter list, parse an ellipsis
to produce a type parameter pack. This replaces
the previous `@_typeSequence` attribute.

hamishknight committed:
[AST] NFC: Refactor GenericTypeParamDecl construction

Add distinct overloads for the parser,
deserialization and code synthesis.

hamishknight committed:
[AST] Remove `@_typeSequence` attribute

This is no longer needed now that we have the
ellipsis spelling.

hamishknight commented:
https://github.com/apple/swift-syntax/pull/948

@swift-ci please test

hamishknight commented:
One interesting case to consider here is whether to allow `<T ...>`, we actually already allow this for regular variadic parameters e.g `func foo(_ x: Int ...)` in addition to `PackExpansionTypeRepr` e.g `(T ...)`. Should we continue to be consistent with regular variadics here, or should we reject this for pack parameters and pack expansion types?

slavapestov commented:
I think allowing a space before `...` is fine and would be consistent with the rest of the language

hamishknight commented:
https://github.com/apple/swift-syntax/pull/948

@swift-ci please test macOS
